axlearn/experiments/text/gpt/param_converter_test.py

```diff
@@ -0,0 +1,174 @@
+
+
+def assert_allclose_in_nested_dict(item1, item2, atol):
+    if isinstance(item1, dict):
+        assert isinstance(item2, dict)
+        assert set(item1.keys()) == set(item2.keys())
+        for k in item1.keys():
+            assert_allclose_in_nested_dict(item1[k], item2[k], atol)
+    else:
+        assert jnp.allclose(item1, item2, atol=atol), f"{np.abs(item1 - item2).max()}"
+
+
+class FujiConvertStateTest(TestCase):
+    def _test_weight_loading(self, fuji_model_name, llama_model_name):
+        trainer_config_map = c4_trainer.named_trainer_configs()
+        trainer_config_fn = trainer_config_map[fuji_model_name]
+        trainer_config = trainer_config_fn()
+        model_config = trainer_config.model
+        model_config.set(name="fuji-test-model")
+        fuji: Model = model_config.instantiate(parent=None)
+        prng_key = jax.random.PRNGKey(0)
+        state = fuji.initialize_parameters_recursively(prng_key=prng_key)
+        config = AutoConfig.from_pretrained(
+            os.path.join(
+                dir_path,
+                "..",
+                "..",
+                "testdata",
+                "axlearn.experiments.text.gpt.param_converter_test",
+                f"{llama_model_name}.json",
+            ),
+            local_files_only=True,
+        )
+        llama = LlamaForCausalLM._from_config(config)  # pylint: disable=W0212
+        llama = llama.eval()
+        ids = jax.random.randint(jax.random.PRNGKey(123), shape=(2, 2), minval=0, maxval=128256)
+        torch_ids = torch.from_numpy(np.asarray(ids))
+        state = parameters_from_llama_3(llama, state)
+        input_batch = {"input_ids": ids}
+        (_, aux), _ = F(
+            fuji,
+            is_training=False,
+            prng_key=jax.random.PRNGKey(123),
+            state=state,
+            inputs={"input_batch": input_batch, "return_aux": True},
+        )
+
+        with torch.no_grad():
+            output = llama(torch_ids)
+
+        fuji_logits = np.asarray(aux["logits"])
+        llama_logits = output.logits.numpy()
+
+        # The difference is caused by the SDPA attention layer. The deeper the larger the error.
+        if fuji_model_name == "fuji-1B-v3":
+            atol = 2e-3
+        elif fuji_model_name == "fuji-3B-v3":
+            atol = 2e-2
+        elif fuji_model_name == "fuji-8B-v3":
+            atol = 2e-1
+        elif fuji_model_name == "fuji-70B-v3":
+            atol = 2.0
+        else:
+            atol = 2e-3
+        assert np.allclose(fuji_logits, llama_logits, atol=atol), (
+            f"{fuji_logits[0,0,:10]} != {llama_logits[0,0,:10]}, "
+            f"{np.abs(fuji_logits - llama_logits).max()}"
+        )
+
+        prng_key, grad_prng_key = jax.random.split(prng_key, 2)
+        fuji_grad = compute_fuji_grad(grad_prng_key, fuji, state, input_batch)
+        llama_grad = compute_llama_grad(llama, torch_ids, state)
+        assert_allclose_in_nested_dict(fuji_grad, llama_grad, atol * 1e-2)
```

Comment:
Does `self.assertNestedAllClose` not work?

Reply:
Done